Av1 improve

feiwan1 opened the pull request.

feiwan1 committed:
avcodec/av1: extend some definitions in spec section 3

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 committed:
cbs_av1: fix incorrect data type

shifted_order_hints is computed by data with int plus data with int.
Switch to int8_t may lose its precision.

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 committed:
avcodec/av1dec: support setup shear process

Defined in spec 7.11.3.6/7.11.3.7.

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 committed:
avcodec/av1dec: add display frame for film grain usage

Make it flexible if decode frame(without apply film grain) and display
frame(applied film grain) both needed when decode film grain clips.

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 committed:
avcodec/vaapi: increase av1 decode pool size

For film grain clip, vaapi_av1 decoder will cache additional 8
surfaces that will be used to store frames which apply film grain.
So increase the pool size by plus 8 to avoid leak of surface.

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 commented:
Did av1 full test locally in TGL, shows no regression.

xuguangxin commented on libavcodec/cbs_av1_syntax_template.c:
```diff
@@ -355,7 +355,7 @@ static int FUNC(set_frame_refs)(CodedBitstreamContext *ctx, RWContext *rw,
     };
     int8_t ref_frame_idx[AV1_REFS_PER_FRAME], used_frame[AV1_NUM_REF_FRAMES];
-    int8_t shifted_order_hints[AV1_NUM_REF_FRAMES];
+    int shifted_order_hints[AV1_NUM_REF_FRAMES];
```
could we have some comment, why we need change this?

feiwan1 replied:
OK, will add it.

xuguangxin commented on libavcodec/av1dec.c:
```diff
@@ -569,6 +569,8 @@ static int get_pixel_format(AVCodecContext *avctx)
 static void av1_frame_unref(AVCodecContext *avctx, AV1Frame *f)
```
maybe a better log is "add file grain support to vaapi"
will this impact dxva?

feiwan1 replied:
This is common part for all av1 hw dec, and the change will happen on some basic struct. So I just want to avoid to mention "vaapi only thing". Anyway, I will think about it, to let the message more acceptable to community.

xuguangxin commented on libavcodec/av1dec.h:
```diff
@@ -33,6 +33,12 @@
 typedef struct AV1Frame {
     ThreadFrame tf;
 
+/**
```
indent

feiwan1 replied:
Will modify, thanks.

xuguangxin commented on libavcodec/av1dec.c:
```diff
@@ -902,9 +936,16 @@ static int set_output_frame(AVCodecContext *avctx, AVFrame *frame,
     int ret;
 
+    /* Use tf_display as output when it is available. */
+    if (s->cur_frame.tf_display.f->buf[0]) {
```
will this impact dxva?

feiwan1 replied:
No impact, just for vaapi.

xuguangxin commented on libavcodec/vaapi_decode.c:
```diff
@@ -577,10 +577,10 @@ static int vaapi_decode_make_config(AVCodecContext *avctx,
         switch (avctx->codec_id) {
         case AV_CODEC_ID_H264:
         case AV_CODEC_ID_HEVC:
+        case AV_CODEC_ID_AV1:
```
will this impact dxva?
why it will leak surfaces?
It's better to move this commit earlier than the previous one, else film grain clips will have an issue with https://github.com/intel-media-ci/ffmpeg/pull/412/commits/db2cf91690316c44cf16e93931979a9f9e0d18a3

feiwan1 replied:
Will modify, thanks.

xuguangxin commented on libavcodec/vaapi_av1.c:
```diff
@@ -63,6 +63,9 @@ static int vaapi_av1_start_frame(AVCodecContext *avctx,
     int err = 0;
     int apply_grain = !(avctx->export_side_data & AV_CODEC_EXPORT_DATA_FILM_GRAIN) && film_grain->apply_grain;
     uint8_t remap_lr_type[4] = {AV1_RESTORE_NONE, AV1_RESTORE_SWITCHABLE, AV1_RESTORE_WIENER, AV1_RESTORE_SGRPROJ};
+    uint8_t segmentation_feature_signed[AV1_SEG_LVL_MAX] = {1, 1, 1, 1, 1, 0, 0, 0};
```
this can be split to multiple commits
each - is a commit

feiwan1 replied:
OK, I will check it. thanks.

xuguangxin commented:
@galinart could you help review and try this on windows?
it will impact av1 dxva too.

thanks

xuguangxin commented:
@uartie , any way to add @galinart to collaborator so we can add him to reviewers.

also, is it possible to disable the "rebase and merge" button for everyone?
we never want merge commits for this repo

thanks

uartie commented:
> @uartie , any way to add @galinart to collaborator so we can add him to reviewers.
> 

Yes, I will add @galinart

> also, is it possible to disable the "rebase and merge" button for everyone?
> we never want merge commits for this repo
> 

No.  GitHub requires at least one option is enabled ☹️ .  We just need to avoid using it.

> thanks

uartie commented:
> > @uartie , any way to add @galinart to collaborator so we can add him to reviewers.
> 
> Yes, I will add @galinart
> 

Collaborator invitation sent to @galinart

uartie commented:
> > also, is it possible to disable the "rebase and merge" button for everyone?
> > we never want merge commits for this repo
> 
> No. GitHub requires at least one option is enabled ☹️ . We just need to avoid using it.
> 

I have enabled "Require pull request reviews before merging"... but max GitHub setting is 6 required reviews.  Also, I enabled "Require conversation resolution before merging".  This should at least keep the "Rebase and Merge" button disabled for most cases.

feiwan1 commented:
@xuguangxin  @galinart, all my change doesn't impact DXVA2/NVDEC,tf_display only used in VAAPI case. But still have question on how DXVA2/NVDEC film grain works? DXVA2/NVDEC does't have interface to support 2 output surface like what VA-API does for each film grain frame. So I guess the only output surface will be applied with film grain. Then according current logic, the output will be used as reference of next frame. But I got from some doc that only reconstruct frame without apply film grain can be used for reference.

xhaihao commented:
> @xuguangxin @galinart, all my change doesn't impact DXVA2/NVDEC,tf_display only used in VAAPI case. But still have question on how DXVA2/NVDEC film grain works? DXVA2/NVDEC does't have interface to support 2 output surface like what VA-API does for each film grain frame. So I guess the only output surface will be applied with film grain. Then according current logic, the output will be used as reference of next frame. But I got from some doc that only reconstruct frame without apply film grain can be used for reference.

Right, the reconstruct is used as reference, the film grain output is for rendering.

feiwan1 committed:
avcodec/av1_vaapi: add gm params valid check

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 committed:
avcodec/av1_vaapi: enable segmentation features

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 committed:
avcodec/av1_vaapi: set display surface when apply film grain

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 committed:
avcodec/av1_vaapi: improve decode quality

- quantizer delta and matrix level specific.
- support loop filter delta.
- support use superres.

Signed-off-by: Fei Wang <fei.w.wang@intel.com>

feiwan1 commented:
HI @xuguangxin  @galinart  any further comments? I am ready to submit this patch to community.

feiwan1 commented:
Hi @a1rwulf , I have submit this patch set to community, could you help comment on it in ffmpeg mail list? Thanks.
https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=4197

feiwan1 commented:
Submitted this PR to ffmpeg community https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=4197, and wait for review. Close this one.

a1rwulf commented:
@feiwan1 thx for the fixes and for submitting it to the ML.
I'll try to see if/how I can help.

rmkimathi commented:
> @feiwan1 thx for the fixes and for submitting it to the ML.
> I'll try to see if/how I can help.

Also waiting for the merge to save me from patching after every git pull ....
